### PR TITLE
h2o ad-hoc tracer

### PR DESCRIPTION
**Pull request by pldubouilh**
Hello

Here's a first proof-of-concept regarding an ad-hoc tracer for h2o based off dtrace probes. The aim was to be cost free when not attached, and still performant under load.

I implemented a small test client for it in python using EBPF/BCC, but any language can be used really. The [script is here](https://gist.github.com/pldubouilh/71de1dd7de2424c6077e5da1f5d16738), and would require [BCC](https://github.com/iovisor/bcc/blob/master/INSTALL.md). I pasted a sample output below.

The PR implements a communication socket to the tracing program. It allows to stop any tracing processing if the tracer program disconnects or crashes, and also some lightweight communication to filter out some events from being processed. This socket port is settable using the config below.

Cheers !

```yaml
listen:
    port: 4321
    tracing: ON
```

```sh
% sudo ./picotracer.py `pgrep h2o`
>> 1553636178731756      36    SSLNew            TLSv1.3 AES256-GCM 0x7ff626ffcac0
>> 1553636178731756      36    NewConnH2         10.0.2.15:9090 10.0.2.15:34200
>> 1553636178731756      36    RxHeader          user-agent curl/7.64.0
>> 1553636178731756      36    RxHeader          accept */*
>> 1553636178731756      36    NewReq            HTTP/2 GET /assets/2.ts
>> 1553636178731756      36    Proxy             [unix:/tmp/h2o.sock]
>> 1553636178731756      36    ProxyTxHdr        user-agent curl/7.64.0
>> 1553636178731756      36    ProxyTxHdr        accept */*
>> 1553636178731756      36    ProxyNewReq       HTTP/1.1 GET /assets/2.ts
>> 1553636178731756      36    ProxyRxHdr        date Tue, 26 Mar 2019 21:36:18 GMT
>> 1553636178731756      36    ProxyRxHdr        content-type video/mp2t
>> 1553636178731756      36    ProxyRxHdr        last-modified Mon, 06 Aug 2018 09:07:10 GMT
>> 1553636178731756      36    ProxyRxHdr        etag "5b680fbe-1a8780"
>> 1553636178731756      36    ProxyRxHdr        accept-ranges bytes
>> 1553636178731756      36    ProxyNewRes       HTTP/1.1 200 len: 1738624
>> 1553636178731756      36    TxHeader          date Tue, 26 Mar 2019 21:36:18 GMT
>> 1553636178731756      36    TxHeader          content-type video/mp2t
>> 1553636178731756      36    TxHeader          last-modified Mon, 06 Aug 2018 09:07:10 GMT
>> 1553636178731756      36    TxHeader          etag "5b680fbe-1a8780"
>> 1553636178731756      36    TxHeader          accept-ranges bytes
>> 1553636178731756      36    NewRes            HTTP/2 200 len: 1738624
```

### CODE REVIEW

File: include/h2o/tracing.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#define H2O_TRACE_RES        ((uint64_t)1<<4)
+#define H2O_TRACE_TXHD       ((uint64_t)1<<5)
+#define H2O_TRACE_RXHD       ((uint64_t)1<<6)
+#define H2O_TRACE_PROXY_REQ  ((uint64_t)1<<7)
+#define H2O_TRACE_PROXY_RES  ((uint64_t)1<<8)
+#define H2O_TRACE_PROXY_TXHD ((uint64_t)1<<9)
+#define H2O_TRACE_PROXY_RXHD ((uint64_t)1<<10)
+
+struct st_h2o_tracing_conn_t {
+    h2o_conn_t super;
```
Comment (kazuho) [inline include/h2o/tracing.h]: I think you do not need to inherit `h2o_conn_t`. `h2o_conn_t` is an abstraction of an HTTP connection (either H1, H2, H3). It exists to help logging accesses to such connections and providing hooks for HTTP version-specific APIs.

Assuming that we are not implementing our own version of HTTP protocol here, what you can do is simply:
* in src/main.c, create a listening socket and accept tracing connections
* pass the established connection to `h2o_tracing_accept`

File: include/h2o/tracing.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    h2o_socket_t *sock;
+
+    /* internal structure */
+    h2o_linklist_t _conns;
+    h2o_req_t req;
+};
```
Comment (kazuho) [inline include/h2o/tracing.h]: Due to the reason stated above, I think you can get rid of the entire structure.

All we need to expose an an API is a function that accepts a socket, and changes the states of the tracer as instructed by the messages that arrive on the socket.

**Comment by deweerdt** [comment]
Should a sample tracing script be shipped with h2o? If so what about including the client linked in the PR? Would that tracing mechanism also work for quicly (cc @janaiyengar ) ?

**Comment by pldubouilh** [comment]
thanks a lot for the review @kazuho :)

I think I covered the 2 comments you raised earlier, but had to slightly amend the current setup on `accept` - feel free to tell me if you would prefer another setup instead :+1:

Also, I changed a bit the way the payloads are carried to the tracing program to make it simpler. I changed the [bcc script](https://gist.github.com/pldubouilh/71de1dd7de2424c6077e5da1f5d16738) accordingly.

File: lib/tracing/tracing.c [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+uint64_t h2o_tracing_now;
+uint64_t should_stop_probe;
+
+#define IS_REMOVE_BIT_SET(x) ((x & H2O_TRACE_REMOVE_PROBE) == H2O_TRACE_REMOVE_PROBE)
+
+static void close_connection(h2o_socket_t *sock)
+{
+    if (should_stop_probe != 0) {
+        // disable selected probe from h2o_tracing_now
```
Comment (janaiyengar) [inline lib/tracing/tracing.c]:
```suggestion
        // disable selected probes from h2o_tracing_now
```

**Comment by kazuho** [comment]
@janaiyengar 
> Specifically, I think we might want to record the entire packet trace in memory (limit the size) and dump it on connection close or when the trace size exceeds a threshold. Basically one trace event per connection. @kazuho, does that make sense?

I think that this is a good question.

IIRC, the concern that lead us to considering doing buffering within h2o then dumping through dprobe is that the overhead of userspace probes could be high, and that eBPF filters cannot accept complex outputs. I anticipate that these assumptions have led us to having the capability of turning on/off the probes from a socket.

I am starting to wonder if the two assumption are connect.

IIIUC, dtrace injects eBPF programs into the process. Therefore, running the programs does not require context switch; it is very lightweight. The probes are NOP instructions when not used, though the overhead of building the arguments for each probe would remain.

eBPF programs have the capability to access any memory location safely, by using [bpf_probe_read](https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#1-bpf_probe_read) and other functions. Therefore, it is possible to pass complex structure to a eBPF filter. The eBPF filter emitting output (by using `perf_submit`) function will cost an IPC (context switch).

Assuming that these understandings are correct, I might argue that the concerns were irrelevant. If that is the case, we might consider:
* Not using a socket to turn on/off the feature. `DTRACE_PROBE` macros can always be invoked, the cost when not used would be negligible.
* Passing complex structures to `DTRACE_PROBE`. Instead of passing method, URL, and the request headers one by one, we can pass all of them at once (e.g., `DTRACE_PROBE7(h2otrace, Request, ts.tv_sec, ts.tv_usec, h2o_socket_get_fd(sock), method, url, headers, num_headers)`). The benefit is that it would reduce the number of context switches by reducing the number of `perf_submit`s, when tracing is used.
* Logging each QUIC event using dtrace. Assuming that we log the events of only a fraction of all the connections, doing IPC for each event would be tolerable. We can use [eBPF maps](https://prototype-kernel.readthedocs.io/en/latest/bpf/ebpf_maps.html) to track the connections being logged. The output can be converted to JSON line in the eBPF program so that existing tooling can be used.

If we can make these changes, the way we use user-space probes will become ordinary. I think that would reduce the cost of maintaining the code in the long run. WDYH?

**Comment by deweerdt** [comment]
> I anticipate that these assumptions have also led us to having the capability of turning on/off the probes from a socket.

What led us to this was the fact that user space probes appear to take a lock, regardless of what the probe does (https://www.mail-archive.com/iovisor-dev@lists.iovisor.org/msg00635.html). This was confirmed experimentally, albeit i haven't tried more recent kernels.
We load tested h2o, and in an environment doing 400k rps, we would see a slowdown of about 15% when the tracing was attached to it -- which i would consider acceptable, given the speed, and the fact that when no probes are attached, it's virtually free.

> If we can make these changes, the way we use user-space probes will become ordinary. I think that would reduce the cost of maintaining the code in the long run. WDYH?

It might be worth redoing these experiments. I wrote the email on uprobes a couple of years ago, the situation might have changed.

**Comment by janaiyengar** [comment]
@kazuho and I discussed this today, and here's where I think we are. @kazuho, please feel free to add more  or disagree with me.

I agree that having two different ways of activating probes (the bitmap and then the probe attachment) seems unnecessary if the cost of always marshalling parameters and running the `DTRACE_PROBE` macro is not high. Some thoughts here:
1. It's worth measuring without the bit-vector to see if we can always run the DTRACE macros. If it's not too expensive, that is preferable.
2. We don't _need_ eBPF maps to track connections that are being logged in general.  They are useful when we want to track arbitrary connections (using Connection ID or IP address based matches, for instance), but for random sampling, quicly could expose state (connection ID for instance) that the eBPF program uses to generate a hash and determine whether to log. 
3. It might still be useful to be able to set state using eBPF. This can be used to enable the marshalling code  when there's a probe attached and similarly enable logging on a connection. In my cursory looking, I've not been able to find a simple off-the-shelf way to do this, but I've asked Subodh at FB who is deeper in eBPF if this is possible.

**Comment by deweerdt** [comment]
> I agree that having two different ways of activating probes (the bitmap and then the probe attachment) seems unnecessary if the cost of always marshalling parameters and running the DTRACE_PROBE macro is not high.

This hinges on the fact that getting the events in user space is cheap when a probe is attached, which hasn't been my experience (hitting the uprobe takes a lock). I'll run some benchmarks to illustrate that point and make sure that this is still valid in recent kernels.

**Comment by kazuho** [comment]
@deweerdt 
> This hinges on the fact that getting the events in user space is cheap when a probe is attached, which hasn't been my experience (hitting the uprobe takes a lock).

I could well be wrong, but IIUC _all_ probes do _not_ become heavy when _any one_ of the probes is attached. Looking at sdt.h, each probe is initially NOP instructions; they are rewritten to instructions that call the eBPF program when a probe is attached. I'd assume that the rewrite happens for only the probes that are being named by the client.

If I am wrong (meaning that actually all probes become heavy when any one is being attached), I'd agree that we might need a socket (or something) that specifies the probes to be activated.

**Comment by deweerdt** [comment]
> I could well be wrong, but IIUC all probes do not become heavy when any one of the probes is attached. Looking at sdt.h, each probe is initially NOP instructions; they are rewritten to instructions that call the eBPF program when a eBPF program is attached. I'd assume that the rewrite happens for only the probes that are being named by the client.

I believe that is correct, but i also believe that doing any kind of connection level tracing will prove too heavy if we don't filter via a selector like the IP or a property of the request.

**Comment by kazuho** [comment]
> I believe that is correct, but i also believe that doing any kind of connection level tracing will prove too heavy if we don't filter via a selector like the IP or a property of the request.

Am I correct to assume that you are considering of adding selector capabilities to the socket interface being proposed, and use userspace eBPF for log shipping? I'd be wary of taking that approach.

IIUC, eBPF provides two features: a) flexibility of defining a selector b) flexibility of shipping the log entries, at the cost of running a VM.

Your argument seems to suggest that we should be using our own selector. That means that the benefit of eBPF would be mostly just _b_. However, if it's just about shipping the log entries, shouldn't we simply emit the log lines from the socket that is used to control the tracer? That'd me much more lightweight.

To clarify, I think the idea of using userspace tracing is great. I would like to be able utilize the full flexibility it provides. Therefore, if it is the case that performance is a concern, I'd suggest tackling the issue by changing the how we define the probes rather than choosing an approach that would reduce the flexibility.

One way that I can think of is to make the probes fat. At the moment, DTRACE_PROBE is called for every request header field. It doesn't necessarily need to be that. As I stated in https://github.com/h2o/h2o/pull/2018#issuecomment-483515896, we can reduce the number of probes being touched to one per HTTP request. I'd hope that that would substantially reduce the overhead, assuming that the eBPF filter program attached to the probe would just ignore the input for most of the time (by checking IP address, etc.).

**Comment by deweerdt** [comment]
> One way that I can think of is to make the probes fat. At the moment, DTRACE_PROBE is called for every request header field. It doesn't necessarily need to be that. As I stated in #2018 (comment), we can reduce the number of probes being touched to one per HTTP request. I'd hope that that would substantially reduce the overhead, assuming that the eBPF filter program attached to the probe would just ignore the input for most of the time (by checking IP address, etc.).

Yes, thanks for stating this, i should have made it clear that i agree that making fat probes is a good idea.

I believe the remaining question whether we want to add a way to optionally call _some_ of the probes. For connection level tracing, such as what's done currently for tracing in quicly, my thinking is we wouldn't be able to build a tcpdump-like utility for QUIC if we don't selectively avoid hitting the probes during a capture. IOW, when we want to troubleshoot a given connection on a live server, the performance impact of hitting the transport level probes for all connections will be to high. Maybe that means that eBPF is a bad choice for that use case, and we should ship the events via a socket? I'm currently thinking that it would be better to use a single tracing system for both cases.

**Comment by deweerdt** [comment]
To give an idea of the kind of performance issue we're talking about, I've written this small test tool: https://github.com/deweerdt/uprobe_bench . We put a probe in a function that's called around 540,000 times per second on my system (4 CPU , then we observe the degradation of the performance as more threads are calling the same function, in a ratio of the non-observed performance to the observed performance:

Threads | Ratio (*)
------------ | -------------
1 |  1.336341
2 | 1.4203
3 | 1.466219
4 | 1.474222

(*) 3rd quantile over 1st quantile performance, see the `summarize` script.

**Comment by janaiyengar** [comment]
> I believe the remaining question whether we want to add a way to optionally call some of the probes. For connection level tracing, such as what's done currently for tracing in quicly, my thinking is we wouldn't be able to build a tcpdump-like utility for QUIC if we don't selectively avoid hitting the probes during a capture. IOW, when we want to troubleshoot a given connection on a live server, the performance impact of hitting the transport level probes for all connections will be to high. Maybe that means that eBPF is a bad choice for that use case, and we should ship the events via a socket? I'm currently thinking that it would be better to use a single tracing system for both cases.

I agree that using a single tracing system is desirable, and I agree that having fat probes is better to minimize overhead. For all current practical uses, I think we can make the packet-tracing probe fat by recording traces in memory and calling DTRACE_PROBE only at a single point per connection: at connection close. It's "live-enough" logging (as against tcpdump's live logging), since an observer will get the trace at the end of the connection, which is sufficient.

Thank you for doing the measurement, @deweerdt ! I'm trying to understand your numbers, so please bear with me while I try. Do I understand correctly that performance is measured as the number of iterations executed in f() per second?

**Comment by deweerdt** [comment]
> Do I understand correctly that performance is measured as the number of iterations executed in f() per second?

@janaiyengar  The performance ratio is measured in terms of 'iterations/second/thread uprobe-not attached' vs 'iterations/second/thread uprobe attached' The higher ratio, the slower the attached version is.

**Comment by kazuho** [comment]
@deweerdt Thank you for sharing the benchmark.

I think that the numbers could be different with USDT (user-level statically defined tracing; the probes that are written in the source code of an application), but I think the benchmark gives us a good ball-park figure (because IIUC USDTs also requires a trap, due to eBPF VM being in the kernel [1]).

> I believe the remaining question whether we want to add a way to optionally call some of the probes. For connection level tracing, such as what's done currently for tracing in quicly, my thinking is we wouldn't be able to build a tcpdump-like utility for QUIC if we don't selectively avoid hitting the probes during a capture. IOW, when we want to troubleshoot a given connection on a live server, the performance impact of hitting the transport level probes for all connections will be to high. Maybe that means that eBPF is a bad choice for that use case, and we should ship the events via a socket? I'm currently thinking that it would be better to use a single tracing system for both cases.

I agree that this is the remaining question. As a solution, my preference is still to use eBPF for choosing the connections to be monitored, because then we do not need to provide a framework for selecting connections in various ways.

@janaiyengar Regarding the question of if we should buffer QUIC events, my preference goes to not. I do not think we would want to buffer all events, then log only a very minor portion of them when each of the connections are close. That means that we'd need to have a selector that selects the connections for which event logging would be turned on. If that is the case, I would assume that using a trap for logging every event would not be an issue, because that'd only happen for the few connections that the selector has chosen.

[1] I missed the fact that it involves a trap (or syscall) in my previous comments, but it nows seems obvious, because there is no userspace VM by default. There exists one though https://github.com/iovisor/ubpf

**Comment by deweerdt** [comment]
> As a solution, my preference is still to use eBPF for choosing the connections to be monitored, because then we do not need to provide a framework for selecting connections in various ways.

That sounds good, thank you for looking into this.

> [1] I missed the fact that it involves a trap (or syscall) in my previous comments, but it nows seems obvious, because there is no userspace VM by default. There exists one though https://github.com/iovisor/ubpf

Interesting, having the instrumentation entirely in user space will make the entire discussing here moot. A quick search didn't look like the infrastructure needed to patch that into a running binary existed though, right?

**Comment by kazuho** [comment]
> Interesting, having the instrumentation entirely in user space will make the entire discussing here moot. A quick search didn't look like the infrastructure needed to patch that into a running binary existed though, right?

Yeah, it looks like it just has the VM. Not the necessary tooling to deal with code injection, etc. Considering that, I think it would be preferable to start with the standard tool-chain (e.g., bpfcc-tools, in-kernel eBPF, DTrace-based API in the application program).

Going back to the discussion of how to write selectors (based on client's IP address, etc.), I am starting to wonder if we can do the marking in the socket receive path (of course it can run eBPF). I mean, something like:
* the eBPF program attached to the incoming socket would mark the SYN<sup>1</sup> packets that needs to be logged, possibly using SO_MARK<sup>2</sup>
* H2O, when accepting a new connection<sup>1</sup>, checks SO_MARK, and if it is set, enables tracing for that connection
* USDT probes can be used for filtering and logging events for the connections that have the tracing enabled

My preference behind this design are:
* The first part (i.e. using eBPF to mark certain connections) can be a performance optimization turned off by default. The default of H2O would be to enable tracing for all of the connections.
* With the first part being optional, we'd be using USDT probes in a pretty standard and cross-platform way.
* Administrators would be given the flexibility to create whatever type of selector and logger they want.

[1] What classifies as a packet that initiates a connection and how it is accepted depends on the protocol (e.g., TCP, QUIC)
[2] I _think_ we can use SO_MARK as a communication path between user program and the eBPF program attached to the socket, though I might be wrong.

**Comment by janaiyengar** [comment]
Thanks @kazuho and @deweerdt for the discussion. 

@kazuho: Yes, it's inadequate to simply have a single probe point per connection, since we only want to record traces in memory for selected connections and not _all_ connections.  To articulate the cost here more precisely, we're talking about a few hundred -- at most 1K -- bytes per packet, so about 1M per connection (if we log a max of 1K packets). That does seem like a fair bit of memory commitment for the common case where it will not be logged.

About the proposed method using SO_MARK, I don't think that works, since for QUIC, multiple connections share the same socket. Consider the following example: The eBPF program reads a packet and sets SO_MARK on the shared socket, then reads a subsequent packet (for a different connection) and unsets the mark. Now H2O reads the first packet, and finds the latest state of SO_MARK, which is unset, erroneously inferring that the first packet's connection need not be traced.

So, I am now questioning whether this is the right approach at all for logging quicly traces, at least packet traces.  I had a chat just now with Subodh at FB, and they use eBPF quite extensively. Subodh uses it for debugging their server, but they do not use eBPF for production logging. My understanding is that they use some sort of IPC for this purpose, and avoid the cost of uprobes for prod logging.

We could get performant logging with just writing to a socket from H2O.  I now wonder what the easiest way to get dynamic control would be without using eBPF.  A strawman would be to have H2O read from the logging socket, where an observer communicates values to use for various filters. We can figure something out here, but this would be a departure from where we wanted to go, with uprobes.

I still think uprobes have value in debugging, but I now wonder if they might be unsuitable for production logging, especially for packet-traces.

**Comment by kazuho** [comment]
@janaiyengar 
> About the proposed method using SO_MARK, I don't think that works, since for QUIC, multiple connections share the same socket. Consider the following example: The eBPF program reads a packet and sets SO_MARK on the shared socket, then reads a subsequent packet (for a different connection) and unsets the mark. Now H2O reads the first packet, and finds the latest state of SO_MARK, which is unset, erroneously inferring that the first packet's connection need not be traced.

I agree that SO_MARK does not work for QUIC. Though I'm pretty sure there'd be a hacky way, for example:
* Use eBPF to add or modify CMSG.
* Use eBPF to route packets to different incoming ports, one port being the one that accepts Initial packets of connections that need to be traced, and the other for accepting all other packets.
* Use eBPF to modify the last bit of the AEAD tag of the Initial packet. 0 would indicate trace-off, 1 would indicate trace-off. We'd lose one bit for AEAD verification, but who cares.

> We could get performant logging with just writing to a socket from H2O. I now wonder what the easiest way to get dynamic control would be without using eBPF. A strawman would be to have H2O read from the logging socket, where an observer communicates values to use for various filters.

Yeah I do not think we'd want to create a vocabulary for naming connections (or packets). I'd rather prefer using code injection, which could be eBPF, mruby proc, lucet, whatever.

**Comment by pldubouilh** [comment]
thanks a lot all for the feedback !

> I agree that this is the remaining question. As a solution, my preference is still to use eBPF for choosing the connections to be monitored, because then we do not need to provide a framework for selecting connections in various ways.

I spent some time trying to find an existing solution for this very issue, but couldn't find much. As you mentioned earlier, the closest is probably [bpf_override_return](https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#1-bpf_override_return), but I think we can only run this for a subset of kprobes. A potential workaround would be     to do a patch to implement `bpf_override_return` for USDT, but I'm not sure of the feasibility regarding this.

> I agree that SO_MARK does not work for QUIC. Though I'm pretty sure there'd be a hacky way

Interesting regarding the socket marking - this could represent an elegant solution for TCP, but I never really thought there would be a possibility with UDP/Quic.

**Comment by deweerdt** [comment]
> [2] I think we can use SO_MARK as a communication path between user program and the eBPF program attached to the socket, though I might be wrong.

We definitely can, and this could be done from either a kprobe or an uprobe, depending on the filter.

> Use eBPF to add or modify CMSG.

A hack that might work would be to set the `IP_RECVTTL` socket option, and we would treat packets receiving the IP TTL CMSG as being traced.

**Comment by kazuho** [comment]
@pldubouilh 
> I spent some time trying to find an existing solution for this very issue, but couldn't find much. As you mentioned earlier, the closest is probably [bpf_override_return](https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#1-bpf_override_return), but I think we can only run this for a subset of kprobes. A potential workaround would be to do a patch to implement `bpf_override_return` for USDT, but I'm not sure of the feasibility regarding this.

Thank you for checking. I think you are correct in pointing out that we cannot use bpf_override_return for our use case, and that we'd not want to start working on modifying eBPF.

@deweerdt 
>> Use eBPF to add or modify CMSG.
> 
> A hack that might work would be to set the `IP_RECVTTL` socket option, and we would treat packets receiving the IP TTL CMSG as being traced.

FWIW I forgot to name in the list the the most straight-forward option: use eBPF maps. An eBPF map is a kernel object that can be used by a eBPF program and the userspace application for sharing data. A map object can be either an array or a hashtable;  userspace application uses syscalls to access the data.

The downside of using eBPF maps is that the signal becomes lossy. Because we would have essentially two queues; one for passing incoming packets and eBPF map for passing the flags to be associated to the packets. Though it could be the case that the loss would be negligible, when we use the 5-tuple or the hash value of the packet as the key of the eBPF map.

Anyways I think the questions regarding this PR are:
* Q1-1. Are we fine with using USDT for HTTP-level tracing?
  * Q1-2. Do we need to determine how to log QUIC-level events before making the decision on Q1-1?
* Q2-1. Are we are fine with using SO_MARK for selecting HTTP/1, HTTP/2 connections?
  * Q2-2. Do we need to determine how to select HTTP/3 connection before making the decision on Q-2-1?

My tentative answers to the questions are: Q1-1: yes, Q1-2: no, Q2-1: I'm not sure, Q2-2: maybe. WDYT?

**Comment by janaiyengar** [comment]
@kazuho 

> FWIW I forgot to name in the list the the most straight-forward option: use eBPF maps.

Indeed -- that occurred to me today, and @deweerdt and I briefly discussed it. To me, that's the least hacky solution, and IMO it should work fine. More further below, but I agree that we do not need to solve the QUIC case before deciding for HTTP-tracing. (Even for the QUIC case, I think we _can_ solve it, I think the question is which mechanism to use.)  If we can set SO_MARK in the eBPF program on the accepted socket, that should work for HTTP/1 and /2. 

So my answers are Q1-1: yes, Q1-2: no, Q2-1: if it can be done, yes, and Q2-2: I don't know.

Going back to using maps: Why do you think using them might be lossy? The design I had in mind had an eBPF program detect Initial packets and simply add the 4-tuple to a map if the connection is "selected". When quicly accepts a new QUIC connection, it checks the map for membership and enables QUIC-level logging if the 4-tuple is present. This map can be quite small since it only needs to hold an entry until quicly sees the incoming connection. This means that the map is queried once per connection, which should be tolerable load.  Is that different than what you had in mind?

**Comment by kazuho** [comment]
> Going back to using maps: Why do you think using them might be lossy? The design I had in mind had an eBPF program detect Initial packets and simply add the 4-tuple to a map if the connection is "selected". When quicly accepts a new QUIC connection, it checks the map for membership and enables QUIC-level logging if the 4-tuple is present. This map can be quite small since it only needs to hold an entry until quicly sees the incoming connection. This means that the map is queried once per connection, which should be tolerable load. Is that different than what you had in mind?

The downside is that you would lose positive signals when the application is not reading entries from the eBPF map fast enough. Also, we need to be careful in deciding how to evict the entries; it would not be a good idea to let H2O do that, because that would mean that it needs to check the eBPF map for every incoming Initial packet (not every new connection), and because there cannot be any losses after the eBPF map is being updated (otherwise the map eventually becomes full).

All that said, I think we'd be OK with using eBPF maps with the key being the 5-tuple, assuming that we would be using either BPF_MAP_TYPE_LRU_HASH or type BPF_MAP_TYPE_LRU_PERCPU_HASH.

**Comment by pldubouilh** [comment]
> `dtrace` command allows us to generate functions (or macros) for emitting information to probes as well as functions for testing if each probe is attached

Very interesting - thanks a lot for this @kazuho, I somehow missed this when reviewing USDT. I'll try amending the PR regarding this :+1:

**Comment by pldubouilh** [comment]
for what it's worth, I did a [small benchmark](https://github.com/pldubouilh/myfirst-usdt) evaluating the cost of testing if the probes were enabled (`TEST_MYPROBE_ENABLED()`). The testbench was running kernel 5.07 with 4 CPU, with no ebpf tracer attached.


for 500k/hits/s

Threads | Ratio 
-- | --
1 | 1.049055
2 | 1.040412
3 | 1.040128
4 | 1.038474

**Comment by deweerdt** [comment]
> Q1-1. Are we fine with using USDT for HTTP-level tracing?
> Q1-2. Do we need to determine how to log QUIC-level events before making the decision on Q1-1?
> Q2-1. Are we are fine with using SO_MARK for selecting HTTP/1, HTTP/2 connections?
> Q2-2. Do we need to determine how to select HTTP/3 connection before making the decision on Q-2-1?
> My tentative answers to the questions are: Q1-1: yes, Q1-2: no, Q2-1: I'm not sure, Q2-2: maybe. WDYT?

I think i'm interested in whether we know today for sure that we don't want to use USDT for QUIC-level events. It sounds like we don't know at this stage, with is fine by me. I agree with your sentiment on the rest of the questions. Thank you!

**Comment by pldubouilh** [comment]
> An eBPF map is a kernel object that can be used by a eBPF program and the userspace application for sharing data

@kazuho cheers for this - I didn't quite realise we could access the map from any program, that's indeed a very nice approach. I implemented this in a [small example](https://gist.github.com/pldubouilh/e930407b9f0c18eae7803ad737917a5b), which:
  * spin an array from an eBPF program
  * pin the array (that makes it system-wide) from the python handler of the eBPF program
  * access the array from the python handler
  * read the array from a C program using the `bpf` syscall

do you think that a similar approach could be implemented in lieu of the socket and messaging I implemented in this PR ? (in addition to your example which test if probes are enabled)

**Comment by kazuho** [comment]
@pldubouilh Thank you for writing down the example! Looks awesome.

> do you think that a similar approach could be implemented in lieu of the socket and messaging I implemented in this PR ? (in addition to your example which test if probes are enabled)

I think that eBPF maps are the way to go for marking the QUIC connections. In addition to marking new connections to be traced, it can potentially be used for marking Initial packets (i.e. SYN packet of QUIC) to which the server needs to respond with RETRYs (i.e. SYN flooding mitigation in QUIC).

I do not have a strong opinion on TCP connections (SO_MARK is another option here), but again, being able to mark new connections for other purposes (e.g., disable TLS 1.3 on certain network paths on which buggy middleboxes exist) might become a feature you need.

In regard to defining the probes in h2o, my preference would be to reuse https://github.com/h2o/picotls/pull/241 like https://github.com/h2o/quicly/pull/166 does. I think I will soon be creating a PR for h2o/h2o that does that, cherry-picking the probe definitions found in this PR. I am in a bit of rush here, because cross-protocol view of what is happening under the hood is essential for HTTP/3 _development_ (for example when you want to know if HTTP/3 prioritization is working as expected on QUIC).

But clearly, the PRs I'm working on do not provide a way of marking certain connections, a feature without which we cannot monitor performance on real _deployments_.

Therefore, I would personally highly appreciate it if you could continue working on how to mark packets attempting to create new connections and to transfer that information to h2o.

@deweerdt @janaiyengar Thoughts?

**Comment by kazuho** [comment]
This PR has been subsumed by #2044 and #2057.

**Pull request closed without merging**